synapse/storage/databases/main/deviceinbox.py

```diff
@@ -664,21 +664,28 @@ def _remove_deleted_devices_from_device_inbox_txn(
             due to a single device having lots of device messages.
             """
 
-            last_stream_id = progress.get("stream_id", 0)
+            txn.execute("SELECT max(stream_id) FROM device_inbox")
+            row = txn.fetchone()
+            if row is None or row[0] is None:
+                max_stream_id = 0
```

Comment:
I'd be tempted just to bail out here. No results means there is nothing to do.
```suggestion
                return 0, True
```